oci/layout/oci_transport.go

```diff
@@ -204,6 +216,67 @@ func (ref ociReference) getManifestDescriptor() (imgspecv1.Descriptor, error) {
 	return imgspecv1.Descriptor{}, ImageNotFoundError{ref}
 }
 
+// Stores an (image) descriptor along with the index it was found in and its parents if any
+// this allows the update of the index when an image is located in a nested (++) index
+type descriptorWrapper struct {
+	descriptor *imgspecv1.Descriptor
+	indexChain []*string //in order of appearence, the first is always be index.json and the nested indexes, last one being the one where the descriptor was found in
+}
+
+// This will return all the descriptors of all the images found in the repository
+// It starts at the index.json and then walks all nested indexes
+// Each image descriptor is returned along with the index it was found, as well as it parents if it is a nested index
+func (ref ociReference) getAllImageDescriptorsInRegistry() ([]*descriptorWrapper, error) {
+	var getImageDescriptorsFromIndex func(indexChain []*string) ([]*descriptorWrapper, error)
+	getImageDescriptorsFromIndex = func(indexChain []*string) ([]*descriptorWrapper, error) {
+		index, err := parseIndex(*indexChain[len(indexChain)-1]) // last item in the index is always the index in which whe are currently working
+		if err != nil {
+			return nil, err
+		}
+
+		descriptors := make([]*descriptorWrapper, 0, len(index.Manifests))
+		for _, manifestDescriptor := range index.Manifests {
+			switch manifestDescriptor.MediaType {
+			case imgspecv1.MediaTypeImageManifest:
+				tmpManifestDescriptor := manifestDescriptor
+				wrapper := descriptorWrapper{&tmpManifestDescriptor, indexChain}
+				descriptors = append(descriptors, &wrapper)
+			case imgspecv1.MediaTypeImageIndex:
+				nestedIndexBlobPath, err := ref.blobPath(manifestDescriptor.Digest, "")
+				if err != nil {
+					return nil, err
+				}
+				// recursively get manifests from this nested index
+				descriptorsFromNestedIndex, err := getImageDescriptorsFromIndex(append(indexChain, &nestedIndexBlobPath))
+				if err != nil {
+					return nil, err
+				}
+				descriptors = append(descriptors, descriptorsFromNestedIndex...)
+			}
+		}
+		return descriptors, nil
+	}
+
+	index := ref.indexPath()
+	indexChain := []*string{&index} // We start the walk at the root: index.json
+	return getImageDescriptorsFromIndex(indexChain)
+}
+
+func (ref ociReference) getManifest(descriptor *imgspecv1.Descriptor) (*imgspecv1.Manifest, error) {
+	//TODO Check if there is shared blob path?
```

Comment:
That’s a good point, actually.

I think one possible answer is that `SystemContext.OCISharedBlobDirPath` did not get set by itself, and the user who set that implicitly opted into a behavior. That might either do exactly what was expected, or be pretty surprising.

Another possible answer is that a single repo just isn’t enough context to safely delete blobs, so we can’t do this correctly — and therefore the image deletion implementation, if it sees `OCISharedBlobDirPath` set, should immediately fail; the user can then either unset this option, or delete the image using… some other tool, if that exists. (And users who have a single-repo `OCISharedBlobDirPath` just can’t use c/image to delete blobs for that repo.)

I think both choices are basically reasonable and I don’t see a strong reason to prefer one over the other, so I think this is up to you.

Immediately failing with a “not supported (yet?)” error would probably be less and simpler code, that might be a relevant factor if we don’t have any stronger reason.

Reply:
I wrote a test and implementation to delete what is found in the local directory and leave what is in the shared folder untouched.

Right now the assumption is that if a blob isn't found, it's likely already gone or in the shared folder. Normally, I would go a step further and check if the blob exists the the shared folder to make a distinction between the two cases and log appropriately, but I'd like your opinion as it adds a few ifs.

Reply:
`blobPath` only returns one value, based on whether `.OCISharedBlobDirPath` is set; so each call of c/image either _only_ uses the local directory, or _only_ uses the shared one. The current implementation does not allow producing/consuming an OCI archive where some blobs are in the local directory, and some others are in the shared directory.

So, it seems to me that “an implementation that deletes what is found in the local directory and leaves what is in the shared folder untouched” is, in effect, a “delete no blobs” implementation. And that’s, actually, another reasonable approach — one I didn’t think of.

If you prefer that behavior, sure, why not.

In that case I just think the code should be clear about that effect — i.e. I’d prefer to remove the code for checks for local blobs in shared-blob situations, so that there are no confusing dead pieces of code.

Reply:
I was indeed unable to find a way to have mixed content directory while tinkering with Skopeo.

In my view, there two considerations here:
1. If someone is using a shared directory, I guess it's because they expect to save space by only storing blobs once. With that in mind, since that shared directory is planned to be used across multiple local directories, so since we have no idea who's using this directory, deleting in there can potentially break other images - that's why I'd rather avoid it.
2. The mixed scenario is rather hypothetical looking at the tools using it right now, is it so far fetched to imagine it could be the case in the future? I agree with the no dead code principle, but in this case it's covered by a test and renders the library more robust.

Saying this, you're the main maintainer here so I'll go with what you prefer. If I have a bit of time, I'll make the changes for the all or nothing approach to compare.

Reply:
Here is a commit with the relevant changes, but I have mixed feelings about this: https://github.com/Pvlerick/image/commit/e0ce6adcab44a60a94ca4ff3e1c303fac5662020

In the end, it amounts to branching out very soon in `DeleteImage`, removing the sharedBobsDir everywhere and always assume that we are in a "local" mode (which is the part I don't like, I tend to be defensive and paranoid with code I write)

Reply:
Hum… I was rather thinking to keep the `sharedBlobsDir` parameter everywhere (to keep the code generic), and just do an

```go
func … DeleteImage() {
   …
   if sharedBlobsDir == "" { // Don't delete in shared blobs mode, we can’t accurately count references
      // …
      // blobsUsedByImage, err = ref.getBlobsUsedInSingleImage(&descriptor, sharedBlobsDir)
      // …
      // … ref.deleteBlobs …
   }
   return ref.deleteReferenceFromIndex(descriptorIndex)
```

but you’re right that keeping a `sharedBlobsDir` parameter while forcing it to be `""` is also redundant/dead code.

---

How about this?
- Keep the actual behavior as it is in this PR
- In `deleteBlobs`, use a rather more prominent comment, something like
  ```go
  // This transport never generates layouts where blobs for an image are both in the local blobs directory
  // and the shared one; it’s either one or the other, depending on how OCISharedBlobDirPath is set.
  //
  // But we can’t correctly compute use counts for OCISharedBlobDirPath (because we don't know what
  // the other layouts sharing that directory are, and we might not even have permission to read them),
  // so we can’t really delete any blobs in that case.
  // Checking the _local_ blobs directory, and deleting blobs from there, doesn't really hurt,
  // in case the layout was created using some other tool or without OCISharedBlobDirPath set, so let's silently
  // check for local blobs (but we should make no noise if the blobs are actually in the shared directory).
  //
  // So, NOTE: the blobPath() call below hard-codes "" even in calls where OCISharedBlobDirPath is set
  for _, digest := range blobsToDelete.Values() {
  		blobPath, err := ref.blobPath(digest, "") 
  ```
- And then ensure that `deleteBlob` really doesn’t produce any warnings if `sharedBlobsDir` is set.

Reply:
Sounds good to me.